Source/AFResult.swift

```diff
@@ -0,0 +1,247 @@
+//
+//  AFResult.swift
+//
+//  Copyright (c) 2014-2019 Alamofire Software Foundation (http://alamofire.org/)
+//
+//  Permission is hereby granted, free of charge, to any person obtaining a copy
+//  of this software and associated documentation files (the "Software"), to deal
+//  in the Software without restriction, including without limitation the rights
+//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+//  copies of the Software, and to permit persons to whom the Software is
+//  furnished to do so, subject to the following conditions:
+//
+//  The above copyright notice and this permission notice shall be included in
+//  all copies or substantial portions of the Software.
+//
+//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+//
+
+import Foundation
+
+public typealias AFResult<T> = Result<T, Error>
```

Comment:
Can the `Error` type constraint be made the more specific `AFError` type? 

I believe Alamofire always vends `AFError`s masked behind `Error`. 
Which currently requires receivers to implicitly know that they should using the `error.asAFError` extension to get the detailed error type.

The switch to explicitly specifying the `Failure` case as `AFError` would eliminate the need for the extension on `Error` and the implicit knowledge of its existence

Reply:
Not really, no, as we don't convert underlying system error to our error type in most situations. It would also require us to wrap all user errors as well, as we don't wrap things like serializer failures in our own error type either. And even if we did wrap everything in `AFError`, the user would still need to extract the underlying associated error as an `Error` and then cast to the type they need. So I'm not sure adding a layer of indirection through our error type would help.

Reply:
This is something we do in our error system built on top of AF @jshier and it's actually super helpful. This way, you can easily categorize the error types you care about. We'd probably need to introduce something like `.user` and `.system` cases into `AFError` that would wrap the underlying error. That would allow us to lock `AFResult` to `AFError` exclusively and would make it easier for clients, even though it makes it more difficult for us in AF. 

Worth considering...

Reply:
Yeah, thinking about it more, the doubly generic `Result` does give us more flexibility here. I just need to think out all of the ramifications of encapsulating all errors.